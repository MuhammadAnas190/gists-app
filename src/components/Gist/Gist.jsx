import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { UserTitle } from '../UserTitle/UserTitle';
import { IconText } from '../IconText/IconText';
import { Date } from '../Date/Date';
import { Files } from '../Files/Files';

export const Gist = ({
  files = [],
  owner: { login, avatar_url: avatarUrl },
  forksUrl,
  comments,
  commentsUrl,
  createdAt,
  updatedAt,
  description,
}) => (
  <CardWrapper>
    <HeaderWrapper>
      <UserTitle name={login} imageSrc={avatarUrl} />
      <HeaderInfoWrapper>
        <IconText iconName="code" text={`${files.length} Files`} />
        <IconText iconName="repo-forked" text="Forks" link={forksUrl} />
        <IconText iconName="comment" text={`${comments} Comments`} link={commentsUrl} />
        <IconText iconName="star" text="Stars" />
      </HeaderInfoWrapper>
    </HeaderWrapper>
    <DateWrapper>
      <Date label="Created at" time={createdAt} />
      <Date label="Updated at" time={updatedAt} />
    </DateWrapper>
    <h4>{description}</h4>
    <Files files={files} />
  </CardWrapper>
);

const CardWrapper = styled.div`
  margin: 0.5rem 0 1rem;
  padding-bottom: 2.5rem;
  display: flex;
  flex-direction: column;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  h4 {
    margin-bottom: 1rem;
  }
`;

const HeaderWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const HeaderInfoWrapper = styled.div`
  display: flex;
  gap: 1rem;
  align-items: center;
`;

const DateWrapper = styled.div`
  display: flex;
  gap: 1rem;
  align-items: center;
  margin: 0.5rem 0 1rem;
`;

Gist.propTypes = {
  files: PropTypes.arrayOf(
    PropTypes.shape({
      filename: PropTypes.string,
      rawUrl: PropTypes.string,
    })
  ).isRequired,
  owner: PropTypes.shape({
    login: PropTypes.string.isRequired,
    avatar_url: PropTypes.string.isRequired,
  }).isRequired,
  forksUrl: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  updatedAt: PropTypes.string,
  commentsUrl: PropTypes.string.isRequired,
  comments: PropTypes.number.isRequired,
  description: PropTypes.string,
};

Gist.defaultProps = {
  description: '',
  updatedAt: null,
};
