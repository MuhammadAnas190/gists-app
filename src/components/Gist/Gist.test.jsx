import { render } from '@testing-library/react';
import { Gist } from './Gist';

const gist = {
  url: 'https://api.github.com/gists/d8c6a231c9c785fa8b2eefb18b1412fd',
  forks_url: 'https://api.github.com/gists/d8c6a231c9c785fa8b2eefb18b1412fd/forks',
  commits_url: 'https://api.github.com/gists/d8c6a231c9c785fa8b2eefb18b1412fd/commits',
  id: 'd8c6a231c9c785fa8b2eefb18b1412fd',
  node_id: 'G_kwDOAH-c9NoAIGQ4YzZhMjMxYzljNzg1ZmE4YjJlZWZiMThiMTQxMmZk',
  git_pull_url: 'https://gist.github.com/d8c6a231c9c785fa8b2eefb18b1412fd.git',
  git_push_url: 'https://gist.github.com/d8c6a231c9c785fa8b2eefb18b1412fd.git',
  html_url: 'https://gist.github.com/bryopsida/d8c6a231c9c785fa8b2eefb18b1412fd',
  files: {
    'install-falcon.yaml': {
      filename: 'install-falcon.yaml',
      type: 'text/x-yaml',
      language: 'YAML',
      raw_url:
        'https://gist.githubusercontent.com/bryopsida/d8c6a231c9c785fa8b2eefb18b1412fd/raw/9b487fe18f1d5b8056b8b1db3b606e52295658d4/install-falcon.yaml',
      size: 871,
    },
  },
  public: true,
  created_at: '2023-09-03T15:00:16Z',
  updated_at: '2023-09-03T15:00:16Z',
  description: 'Ansible playbook to install Falco on Debian systems',
  comments: 0,
  user: null,
  comments_url: 'https://api.github.com/gists/d8c6a231c9c785fa8b2eefb18b1412fd/comments',
  owner: {
    login: 'bryopsida',
    id: 8363252,
    node_id: 'MDQ6VXNlcjgzNjMyNTI=',
    avatar_url: 'https://avatars.githubusercontent.com/u/8363252?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/bryopsida',
    html_url: 'https://github.com/bryopsida',
    followers_url: 'https://api.github.com/users/bryopsida/followers',
    following_url: 'https://api.github.com/users/bryopsida/following{/other_user}',
    gists_url: 'https://api.github.com/users/bryopsida/gists{/gist_id}',
    starred_url: 'https://api.github.com/users/bryopsida/starred{/owner}{/repo}',
    subscriptions_url: 'https://api.github.com/users/bryopsida/subscriptions',
    organizations_url: 'https://api.github.com/users/bryopsida/orgs',
    repos_url: 'https://api.github.com/users/bryopsida/repos',
    events_url: 'https://api.github.com/users/bryopsida/events{/privacy}',
    received_events_url: 'https://api.github.com/users/bryopsida/received_events',
    type: 'User',
    site_admin: false,
  },
  truncated: false,
};

describe('Gist Component', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Gist {...gist} files={Object.values(gist.files)} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
