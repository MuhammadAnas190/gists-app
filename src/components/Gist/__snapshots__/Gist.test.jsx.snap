// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gist Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsFSXq cCMyxZ"
  >
    <div
      class="sc-kAyceB hLXIir"
    >
      <div
        class="sc-aXZVg cWrJRO"
      >
        <img
          class="sc-gEvEer cjclob"
          src="https://avatars.githubusercontent.com/u/8363252?v=4"
        />
        <p>
          bryopsida
        </p>
      </div>
      <div
        class="sc-imWYAI dQcJWG"
      >
        <div
          class="sc-eqUAAy ecMSCB"
        >
          <span
            class="octicon octicon-code"
          />
          <p>
            1 Files
          </p>
        </div>
        <div
          class="sc-eqUAAy ecMSCB"
        >
          <span
            class="octicon octicon-repo-forked"
          />
          <p>
            Forks
          </p>
        </div>
        <div
          class="sc-eqUAAy ecMSCB"
        >
          <span
            class="octicon octicon-comment"
          />
          <p>
            0 Comments
          </p>
        </div>
        <div
          class="sc-eqUAAy ecMSCB"
        >
          <span
            class="octicon octicon-star"
          />
          <p>
            Stars
          </p>
        </div>
      </div>
    </div>
    <div
      class="sc-jXbUNg gVHlFf"
    >
      <div
        class="sc-dcJsrY RPpny"
      >
        <span>
          Created at:
        </span>
        <span>
          N/A
        </span>
      </div>
      <div
        class="sc-dcJsrY RPpny"
      >
        <span>
          Updated at:
        </span>
        <span>
          N/A
        </span>
      </div>
    </div>
    <h4>
      Ansible playbook to install Falco on Debian systems
    </h4>
    <div
      class="sc-iGgWBj gXEntF"
    >
      <div
        class="sc-eqUAAy ecMSCB"
      >
        <span
          class="octicon octicon-file"
        />
        <a
          class="sc-fqkvVR jDLZtY"
          href="https://gist.githubusercontent.com/bryopsida/d8c6a231c9c785fa8b2eefb18b1412fd/raw/9b487fe18f1d5b8056b8b1db3b606e52295658d4/install-falcon.yaml"
          target="_blank"
        >
          install-falcon.yaml
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
